ADALiOS/ADALiOS/ADKeychainItem.m

```diff
@@ -0,0 +1,366 @@
+{
+    for (ADKeychainToken* token in _accessTokens)
+    {
+        // Ignore anything in the keychain item that's not the correct class.
+        if (![token isKindOfClass:[ADKeychainToken class]])
+        {
+            continue;
+        }
+        
+        if ([scopes isSubsetOfSet:token.scopes])
+        {
+            return token;
+        }
+    }
+    
+    return nil;
+}
+
+- (void)addAccessTokenWithScopes:(NSSet*)scopes
+                          toItem:(ADTokenCacheStoreItem*)item
+{
+    item.refreshToken = _refreshToken;
+    
+    ADKeychainToken* token = [self tokenWithScopes:scopes];
+    [token addToTokenItem:item];
+}
+
+- (void)removeIntersectingTokens:(NSSet*)scopes
+{
+    NSMutableIndexSet* toRemove = [NSMutableIndexSet indexSet];
+    
+    NSUInteger cTokens = [_accessTokens count];
+    
+    for (NSUInteger i = 0; i < cTokens; i++)
+    {
+        ADKeychainToken* token = _accessTokens[i];
+        
+        // Ignore anything in the keychain item that's not the correct class.
+        if (![token isKindOfClass:[ADKeychainToken class]])
+        {
+            continue;
+        }
+        
+        if ([scopes intersectsSet:token.scopes])
+        {
+            [toRemove addIndex:i];
+        }
+    }
+    
+    [_accessTokens removeObjectsAtIndexes:toRemove];
+}
+
+- (void)updateToTokenItem:(ADTokenCacheStoreItem*)item
+{
+    _refreshToken = item.refreshToken;
+    
+    NSSet* scopes = item.scopes;
+    [self removeIntersectingTokens:scopes];
+    
+    ADKeychainToken* token = [ADKeychainToken new];
+    [token updateToTokenItem:item];
+    [_accessTokens addObject:token];
+}
+
+@end
+
+@interface ADKeychainItem ()
+
+@property NSString* authority;
+@property NSString* clientId;
+@property NSData* sessionKey;
+@property ADProfileInfo* profileInfo;
+
+@end
+
+
+@implementation ADKeychainItem
+{
+    NSMutableDictionary* _policies;
+}
+
+
+
++ (ADKeychainItem*)itemForData:(NSData *)data
+{
+    if (!data)
+    {
+        return nil;
+    }
+    
+    id item = [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    if (![item isKindOfClass:[ADKeychainItem class]])
+    {
+        return nil;
+    }
+    
+    return item;
+}
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _policies = [NSMutableDictionary new];
+    
+    return self;
+}
+
+- (id)initWithCoder:(NSCoder*)aDecoder
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _authority = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"authority"];
+    _clientId = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"clientId"];
+    _sessionKey = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"sessionKey"];
+    _profileInfo = [aDecoder decodeObjectOfClass:[ADProfileInfo class] forKey:@"profileInfo"];
+    
+    _policies = [aDecoder decodeObjectOfClass:[NSArray class] forKey:@"policies"];
```

Comment:
Let's step back. Policies should not be anything but a nsstring. There will only be one policy per token. Why is this an array?

Sent from Outlook<http://aka.ms/Ox5hz3>




On Wed, Sep 16, 2015 at 4:21 PM -0700, "Ryan Pangrle" <notifications@github.com<mailto:notifications@github.com>> wrote:


In ADALiOS/ADALiOS/ADKeychainItem.m<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzureAD%2fazure-activedirectory-library-for-objc%2fpull%2f378%23discussion_r39697317&data=01%7c01%7cKanishk.Panwar%40microsoft.com%7c73b64227fc124720aa8b08d2beed81ad%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=R%2bYeLxHJHryCYDd0cnka%2b8wes92WMjLk6cYsOvDY8TU%3d>:

> +    return self;
> +}
> +
> +- (id)initWithCoder:(NSCoder*)aDecoder
> +{
> +    if (!(self = [super init]))
> +    {
> +        return nil;
> +    }
> +
> +    _authority = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"authority"];
> +    _clientId = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"clientId"];
> +    _sessionKey = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"sessionKey"];
> +    _profileInfo = [aDecoder decodeObjectOfClass:[ADProfileInfo class] forKey:@"profileInfo"];
> +
> +    _policies = [aDecoder decodeObjectOfClass:[NSArray class] forKey:@"policies"];


This should be NSDictionary, not either array or set, as this is the keychain data object.

-
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzureAD%2fazure-activedirectory-library-for-objc%2fpull%2f378%2ffiles%23r39697317&data=01%7c01%7cKanishk.Panwar%40microsoft.com%7c73b64227fc124720aa8b08d2beed81ad%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=QxIZm1Yunt%2bkrmRR%2fNpt34TGxCw7vadM1A01Ig3R5Hc%3d>.